tests/unit/sharded_copying_manager_tests/copying_manager_worker_test.py

```diff
@@ -0,0 +1,695 @@
+
+        return worker
+
+    def _init_worker_instance(
+        self,
+        log_files_number=1,
+        auto_start=True,
+        add_processors=True,
+        use_pipelining=False,
+        disable_flow_control=False,
+    ):  # type: (int, bool, bool, bool) -> Tuple[Tuple[TestableLogFile, ...], TestableCopyingManagerThreadedWorker]
+
+        config_data = {
+            "api_keys": [{"workers": 1}],
+            "use_multiprocessing_workers": self.use_multiprocessing_workers,
+            "disable_send_requests": True,
+        }
+
+        self._create_config(
+            log_files_number=log_files_number,
+            use_pipelining=use_pipelining,
+            config_data=config_data,
+        )
+
+        self._config_builder.config.disable_flow_control = disable_flow_control
+
+        self._instance = self._create_worker(
+            self._config_builder.config,
+            self._config_builder.config.api_key_configs[0],
+            "1",
+        )
+
+        if add_processors:
+            for test_file in self._config_builder.log_files.values():
+                self._spawn_single_log_processor(test_file)
+
+        if auto_start:
+            self._instance.start_worker()
+
+        test_files = tuple(self._config_builder.log_files.values())
+
+        return test_files, self._instance
+
+    def _spawn_single_log_processor(
+        self, log_file, checkpoints=None, copy_at_index_zero=False,
+    ):
+        # type: (TestableLogFile, Optional[Dict], bool)-> LogFileProcessor
+        log_config = self._config_builder.get_log_config(log_file)
+
+        matcher = LogMatcher(self._config_builder.config, log_config)
+        if checkpoints is None:
+            checkpoints = {}
+
+        processors = matcher.find_matches(
+            existing_processors=[],
+            previous_state=checkpoints,
+            copy_at_index_zero=copy_at_index_zero,
+            create_log_processor=self._instance.create_and_schedule_new_log_processor,
+        )
+
+        return processors[0]
+
+
+class TestsRunning(CopyingManagerWorkerTest):
+    @skip_on_multiprocess_workers
+    def test_wait_copying(self):
+
+        run_released = threading.Event()
+
+        _, worker = self._init_worker_instance(1, auto_start=False)
+
+        original_run = worker.run
+
+        # create mock run method that can wait for signal to start
+        def delayed_run():
+            # we need to set this event to resume worker.
+            run_released.wait()
+            return original_run()
+
+        with mock.patch.object(worker, "run", wraps=delayed_run):
+            executor = ThreadPoolExecutor()
+
+            worker.start_worker(stop_at=None)
+
+            # wait for copying start in separate thread
+
+            wait_future = executor.submit(worker.wait_for_copying_to_begin)
+
+            # check for waiting is finished, it must not be finished until event is not set.
+            assert not wait_future.done()
+            time.sleep(0.1)
```

Comment:
Why sleep here like this?  It always worries me to see lines like this.  Normally the right thing to do is to sleep until some condition becomes true.